src/wasm/wasm-type.cpp

```diff
@@ -2872,6 +2954,60 @@ std::vector<HeapType> buildEquirecursive(TypeBuilder& builder) {
   return heapTypes;
 }
 
+void validateSubTyping(const std::vector<HeapType>& heapTypes) {
+  // Ensure there are no cycles in the subtype graph.
```

Comment:
maybe add an assert that the type system is nominal? (or, the name could include "nominal" in it perhaps)